File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week3/Programming_Assignments/6_maximum_number_of_prizes/different_summands.py
# Uses python3
import sys

def optimal_summands(n):
    summands = []
    i = 0
    while n > 0:
        i += 1
        if n - i <= i:
            i = n
        summands.append(i)
        n -= i
    return summands

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    summands = optimal_summands(n)
    print(len(summands))
    for x in summands:
        print(x, end=' ')

File: Data_Structures_And_Algorithms-UC_SanDiego/1-Algorithmic_Toolbox/Week3/Programming_Assignments/6_maximum_number_of_prizes/different_summands.cpp
#include <iostream>
#include <vector>

using std::vector;

vector<int> optimal_summands(int n) {
  vector<int> summands;
  int i = 0;
  while (n>0){
    i += 1;
    if (n-i <= i){ i = n; }
    summands.push_back(i);
    n = n - i;
  }
  return summands;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> summands = optimal_summands(n);
  std::cout << summands.size() << '\n';
  for (size_t i = 0; i < summands.size(); ++i) {
    std::cout << summands[i] << ' ';
  }
}
